Bug fix: Not setting action after adding to table.

bdevierno1 opened the pull request:
Test flow directory NF bug fix.

<!-- Add detailed description and provide running instructions -->
## Summary:

The NF was not setting the meta action after adding a new flow to the table.

**Usage:** 
As described in readme.

<!-- Check list of the things this PR accomplishes -->
| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          | <!-- Provide a list of issues --> 
| Breaking API changes     |
| Internal API changes     |
| Usability improvements   |  
| Bug fixes                |
| New functionality        |
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 


<!-- If the pr has any dependencies or merge quirks note them here -->
## Merging notes:
 - Dependencies: None  

**TODO before merging :**
 - [ ] PR is ready for review

I hope to be making changes to this in the coming few days to add an example for service chaining. As of now this resolves issue #247 

<!-- What you did to test the PR, what needs to be done -->
## Test Plan:

Run NF with Pktgen. Test that NF does not crash as before. 

<!-- Notes about what you think should be reviewed, any specific hacks in this PR -->
## Review: 
Anyone.

onvm commented:
> In response to PR creation

## CI Message

Your results will arrive shortly

onvm commented:
> In response to PR creation

## CI Message

Error: Failed to parse Pktgen stats

bdevierno1 commented:
@onvm Hello HAL, do you read me?

onvm commented:
> @onvm Hello HAL, do you read me? 

## CI Message

Your results will arrive shortly

onvm commented:
> @onvm Hello HAL, do you read me? 

## CI Message

Error: Failed to fetch results from nimbnode23

bdevierno1 commented on examples/test_flow_dir/test_flow_dir.c:
```diff
@@ -155,19 +155,27 @@ static int
                __attribute__((unused)) struct onvm_nf_local_ctx *nf_local_ctx) {
         static uint32_t counter = 0;
-        struct onvm_flow_entry *flow_entry = NULL;
+        struct onvm_flow_entry *flow_entry;
```
Initializing this to NULL resulted in a seg fault when creating the chain flow_entry->sc.

twood02 replied:
It seems unlikely that removing the NULL is actually fixing the problem - it may just be masking the error.  It is better to initialize it to NULL and then figure out why it is being accessed before being initialized properly.

bdevierno1 commented on examples/test_flow_dir/test_flow_dir.c:
```diff
@@ -155,19 +155,27 @@ static int
+        struct onvm_ft_ipv4_5tuple key;
+        memset(&key, 0, sizeof(struct onvm_ft_ipv4_5tuple));
+        onvm_ft_fill_key(&key, pkt);
+        ret = onvm_ft_lookup_key(sdn_ft, &key, (char **)flow_entry);
```
I noticed that using the lookup pkt in the flow director api either resulted in a seg fault or did not perform a correct lookup or add to the table.

bdevierno1 commented on examples/test_flow_dir/test_flow_dir.c:
```diff
@@ -176,7 +184,8 @@ packet_handler(struct rte_mbuf *pkt, struct onvm_pkt_meta *meta,
                 onvm_sc_append_entry(flow_entry->sc, ONVM_NF_ACTION_TONF, destination);
-                // onvm_sc_print(flow_entry->sc);
+                meta->action = ONVM_NF_ACTION_TONF;
+                meta->destination = destination;
```
Previously the meta action was not set which resulted in seg fault.

bdevierno1 commented on onvm/onvm_nflib/onvm_flow_table.h:
```diff
@@ -71,8 +71,8 @@ struct onvm_ft {
-        uint16_t src_port;
-        uint16_t dst_port;
+        uint32_t src_port;
+        uint32_t dst_port;
```
When running pktgen it default to a 5000+ size port number. This resulted in a seg fault. Im not to sure why this  was the case as it would only occur if adding a flow to the the shared flow table and not for example in the flow tracker which initializes a new flow table in the secondary process. However, increasing the size of the tuple ameliorated this occurence.

twood02 replied:
what do you mean by `5000+ size port number`?  Increasing the size of a uint generally shouldn't fix a seg fault, but it suggests somewhere we might be allocating or memcpying into here something that is the wrong size.  In our other `key` datastructures do we use a 16 or 32bit port?  The IP specification says it should be 16 bit so it would be better to keep the struct that way.

bdevierno1 replied:
When creating a key with a port number greater that 1000+ a seg fault would occur. I printed the key that is created after a packet is arrived using onvm_ft_print_key, setting pktgen to have destination port of 1. However, I then noticed it would print the integer '256' after creating the key. In my last commit I converted the value to big endian. The correct port values are now being set and the seg fault no longer occurs.

bdevierno1 commented on onvm/onvm_nflib/onvm_flow_table.h:
```diff
@@ -148,7 +148,6 @@ onvm_ft_fill_key(struct onvm_ft_ipv4_5tuple *key, struct rte_mbuf *pkt) {
                 return -EPROTONOSUPPORT;
         }
         ipv4_hdr = onvm_pkt_ipv4_hdr(pkt);
-        memset(key, 0, sizeof(struct onvm_ft_ipv4_5tuple));
```
When filling in the key, passing in a pointer within the pkt handler from a secondary process resulted in a seg fault.

bdevierno1 commented on onvm/onvm_nflib/onvm_flow_table.h:
```diff
@@ -148,18 +148,17 @@ onvm_ft_fill_key(struct onvm_ft_ipv4_5tuple *key, struct rte_mbuf *pkt) {
-                key->src_port = tcp_hdr->src_port;
-                key->dst_port = tcp_hdr->dst_port;
+                key->src_port = rte_cpu_to_be_16(tcp_hdr->src_port);
+                key->dst_port = rte_cpu_to_be_16(tcp_hdr->dst_port);
```
Convert to big endian. DPDK has examples of using these here https://doc.dpdk.org/api/examples_2performance-thread_2l3fwd-thread_2main_8c-example.html.

bdevierno1 commented:
In my last commit I have added a new feature/example populating service chains to the flow table #60 . You can then use this to test the onvm flow dir API by starting up up to four test_flow_dir NFs. Another way is to run this NF to populate the flow table then subsequently use the simple forward NF;  but first modifying the packet handler to use the action next instead. 

I have made further changes to the flow table api as it was printing the keys I had pre-defined backwards.

I have reverted the flow_entry variable to initialize to NULL. However, I had to allocate memory within the packet handler which will probably be slower. I will update this soon as there is probably a better way to do this.

dennisafa commented on examples/test_flow_dir/test_flow_dir.c:
```diff
@@ -151,32 +160,96 @@ do_stats_display(struct rte_mbuf *pkt) {
+        for (uint32_t i = 0; i < num_keys; i++) {
+                struct onvm_ft_ipv4_5tuple key;
+                memset(&key, 0, sizeof(struct onvm_ft_ipv4_5tuple));
+                key = keys[i].key;
```
Why do we memset to 0 before reassigning?

bdevierno1 replied:
Updated. Thanks!

dennisafa commented on examples/test_flow_dir/test_flow_dir.c:
```diff
@@ -151,32 +160,96 @@ do_stats_display(struct rte_mbuf *pkt) {
-        ret = onvm_flow_dir_get_pkt(pkt, &flow_entry);
+        struct onvm_ft_ipv4_5tuple key;
+        memset(&key, 0, sizeof(struct onvm_ft_ipv4_5tuple));
+        onvm_ft_fill_key(&key, pkt);
```
Fill the key with the packet once you hit the else statement on line 242, since we don't want to do it needlessly.

bdevierno1 replied:
I believe I will need to fill the key here to perform the lookup.

twood02 commented on examples/test_flow_dir/test_flow_dir.c:
```diff
@@ -82,6 +86,8 @@ usage(const char *progname) {
         printf("Flags:\n");
         printf(" - `-d <dst>`: destination service ID to foward to\n");
         printf(" - `-p <print_delay>`: number of packets between each print, e.g. `-p 1` prints every packets.\n");
+        printf(" - `-f : Enable populating sdn flow table. \n");
```
change this to `-f : Prepopulate sample flow table rules`

twood02 replied:
make it `-s`

twood02 commented on examples/load_balancer/load_balancer.c:
```diff
@@ -456,6 +456,7 @@ table_lookup_entry(struct rte_mbuf *pkt, struct flow_info **flow) {
                 return -1;
         }
 
+        memset(&key, 0, sizeof(struct onvm_ft_ipv4_5tuple));
```
try putting these memsets back inside the fill_key functions and see if there is still a problem.

twood02 commented on examples/test_flow_dir/README.md:
```diff
@@ -3,6 +3,8 @@ Flow Director Test
 
 This NF demonstrates how to use ONVM's Flow Director. When a packet arrives the NF checks whether it is from a flow that already has a service chain rule. If not, it creates a new rule so the packet will be sent to the destination NF indicated on the command line. Packets that match a rule are processed with the ONVM_NF_ACTION_NEXT action.
 
+This NF demonstrates how to populate the SDN flow table as well as how to create service chain rules for specific flows. To enable this functionality use the -f flag.
```
Rename "SDN flow table" to "Manager's flow table"

twood02 commented on examples/test_flow_dir/README.md:
```diff
@@ -24,6 +26,7 @@ App Specific Arguments
 --
   - `-d <dst>`: destination service ID to foward to
   - `-p <print_delay>`: number of packets between each print, e.g. `-p 1` prints every packets.
+  - `f`: populate ONVM's flow director table with predefined flows.
```
update comment as below

twood02 commented on examples/test_flow_dir/test_flow_dir.c:
```diff
@@ -151,32 +160,102 @@ do_stats_display(struct rte_mbuf *pkt) {
+ */
+
+static void
+populate_ipv4_flow_table(void) {
```
rename to `populate_sample_ipv4....` and add comment explaining that users can define their flow rules here if they want

bdevierno1 commented:
![image](https://user-images.githubusercontent.com/54540257/88645604-e0301080-d0f6-11ea-912c-5d9b0a1b2018.png)
This is ready for review. If users set destination to 255 it will use the next action.

bdevierno1 commented:
@onvm Hello HAL, do you read me?

onvm commented:
> @onvm Hello HAL, do you read me

## CI Message

Your results will arrive shortly

onvm commented:
> @onvm Hello HAL, do you read me

## CI Message

Your results will arrive shortly

kevindweb commented:
@onvm is it really slow?

onvm commented:
> @onvm is it really slow?

## CI Message

Your results will arrive shortly

koolzz commented on onvm/onvm_nflib/onvm_pkt_common.c:
```diff
@@ -102,7 +102,7 @@ onvm_pkt_process_tx_batch(struct queue_mgr *tx_mgr, struct rte_mbuf *pkts[], uin
                         nf->stats.act_drop++;
                         nf->stats.tx += !onvm_pkt_drop(pkts[i]);
-                } else if (meta->action == ONVM_NF_ACTION_NEXT) {
+                } else if (meta->action == ONVM_NF_ACTION_NEXT || meta->destination == 255) {
```
No magic numbers please, what is this 255? Why do we need it? 
Can't the packet just set action ONVM_NF_ACTION_NEXT?

twood02 replied:
Having a special number is purposeful since we want to use this to indicate the next action even if the NF wasn't designed for that purpose (eg this lets us make the SimpleFwd NF use the "next action" behavior if you set 255 at the command line as the destination ID). 

But you are right that it shouldn't be totally magic -- @bdevierno1 you should #define a constant somewhere for this instead of using 255 directly here.

koolzz replied:
Sure, feels like a lazy way to change the action(we could just bake in the same `if meta->destination -> meta->action = ONVM_NF_ACTION_NEXT` into the NF itself) but if people find it easier lets define a global const.

koolzz replied:
@bdevierno1 Also if we're doing this, after #defining the constant (probably in common.h) do a onvm mgr startup check to confirm that your const number is more than MAX_NFS. Otherwise if someone has 512 NFS configured they will be suprised that forwarding to 256 doesn't work

bdevierno1 replied:
This line was resulting in the perf drop with speed tester. I couldn't find another way to optimize this besides moving the ONVM_NF_ACTION_TONF if statement above the one for the NEXT action on line 105. But I think this would feel like it would be an over-optimization for speed tester? I'll update to use the global const for now. I could also update all the sample NFs to use a flag instead to indicate to use the next action as well if we decide that's better.

bdevierno1 commented:
@onvm Test please.

onvm commented:
> @onvm Test please.

## CI Message

Your results will arrive shortly

bdevierno1 commented:
Ran two tests using c220g1 nodes in Wisconsin site. Noticed about a 2% performance drop. 
![Screen Shot 2020-08-03 at 3 59 31 PM](https://user-images.githubusercontent.com/54540257/89161188-db5ed700-d5a4-11ea-8e9f-6ac39c444340.png)
Without setting 255 as next action.
![Screen Shot 2020-08-03 at 4 01 07 PM](https://user-images.githubusercontent.com/54540257/89161198-dd289a80-d5a4-11ea-8d8f-6f72d8ef3113.png)

kevindweb commented on onvm/onvm_nflib/onvm_pkt_common.c:
```diff
@@ -102,9 +102,8 @@ onvm_pkt_process_tx_batch(struct queue_mgr *tx_mgr, struct rte_mbuf *pkts[], uin
-                } else if (meta->action == ONVM_NF_ACTION_NEXT) {
-                        /* TODO: Here we drop the packet : there will be a flow table
-                        in the future to know what to do with the packet next */
+                } else if (meta->action == ONVM_NF_ACTION_NEXT || meta->destination == ACTION_NEXT_DEST_ID) {
```
@bdevierno1 setting 255 as the next action would definitely decrease performance, but I think that (from CI) we see that just having to process this or `||` statement in the condition slows down other NFs. Because we know that speed_tester would not have set the `meta->destination` to 255. I think that's the cause of the ~4% performance drop from CI

bdevierno1 replied:
Yeah this is correct. Unfortunately, I tried different ways to optimize this but did not see any substantial improvements. By removing the OR I will still have to set another conditional which will have to be placed before this statement so as to maintain the same logic. Perhaps there is another way to do this?

dennisafa replied:
Instead of doing meta->action == 255, can we define a new action that denotes this is a packet meant for flow rule lookup?

kevindweb commented on examples/test_flow_dir/test_flow_dir.c:
```diff
@@ -214,9 +298,16 @@ main(int argc, char *argv[]) {
         onvm_nflib_run(nf_local_ctx);
 
         onvm_nflib_stop(nf_local_ctx);
+        flow_entry = NULL;
+        rte_free(flow_entry);
```
These as well I think.

bdevierno1 replied:
This should be resolved. I did this as I was DPDK stated I was accessing invalid memory. This solution was misguided,  should be resolved. Thanks!

kevindweb commented on examples/test_flow_dir/test_flow_dir.c:
```diff
@@ -151,32 +163,103 @@ do_stats_display(struct rte_mbuf *pkt) {
+                        }
+                }
+        }
+        flow_entry = NULL;
+        rte_free(flow_entry);
```
Am I missing something here or should these be flipped?

bdevierno1 replied:
Thank You.
